### PR TITLE
[#2373] Support datetime for since and until repo-config.csv

### LINKED ISSUE #2373: Support datetime for since and until repo-config.csv
Currently, we have #2369 to support datetime for since and until through CLI. Due to urgency for the CLI functionality to be merged, let's move support for datetime in repo-config.csv as another separate issue, as it may contain significant test and system test changes, as well as User Guide updates.

### PR DESCRIPTION
**Pull request by gabriellegtw**
Fixes #2373 

### Proposed commit message
```
Support datetime (previously just date) for since and until in repo-config.csv

3 supported formats for since and until dates: 
1. DD/MM/YYYY
2. DD/MM/YYYY'T'HH:mm
3. DD/MM/YYYY'T'HH:mm:ss

Behavior: For since date, if hour/minute/seconds not supplied then
default to 0 for each. For until date, if hour/minute/seconds not
supplied then default to 23, 59, 59 respectively.
```

### CODE REVIEW

File: src/main/java/reposense/parser/RepoConfigCsvParser.java [hunk 1]
```diff
@@ -240,10 +239,25 @@ private void checkValidDatesWithCli(LocalDateTime date) throws InvalidDatesExcep
     private LocalDateTime extractCsvSinceDate(CSVRecord record) throws InvalidDatesException {
         String sinceDateStr = get(record, SINCE_HEADER);
         boolean hasUpdatedSinceDateTime = !sinceDateStr.isEmpty();
+
```
Comment (CYX22222003) [inline src/main/java/reposense/parser/RepoConfigCsvParser.java:242]: Good use of `DateTimeFormatterBuilder` to relax the constraint on the format of date time input.

File: src/test/java/reposense/parser/RepoConfigParserTest.java [hunk 2]
```diff
@@ -66,12 +66,30 @@ public class RepoConfigParserTest {
             "RepoConfigParserTest/repoconfig_ignoreFileSizeLimit_test.csv");
     private static final Path REPO_CONFIG_INVALID_OVERRIDEN_DATE = loadResource(RepoConfigParserTest.class,
             "CsvParserTest/repocsvconfig_invalidOverriddenDate_test.csv");
+    private static final Path REPO_CONFIG_INVALID_OVERRIDEN_DATE_WITH_TIME = loadResource(RepoConfigParserTest.class,
```
Comment (CYX22222003) [inline src/test/java/reposense/parser/RepoConfigParserTest.java:69]: Good to include test cases for different forms of valid CSV date formats.

File: docs/ug/configFiles.md [hunk 3]
```diff
@@ -31,22 +31,22 @@ Given below are the details of the various config files used by RepoSense.
...
+| Since Date                            | Enter since date (with optional time specification) in one of the following formats: <br>• `DD/MM/YYYY`<br>• `DD/MM/YYYY'T'HH:mm`<br>• `DD/MM/YYYY'T'HH:mm:ss`<br><br> This signifies the start date of analysis. If the field is ignored, the date will be set to the default one or the date indicated in CLI flags <br><br> If hours/ minutes/ seconds are not provided, each will default to `00` |
+| Until Date                            | Enter until date (with optional time specification) in one of the following formats: <br>• `DD/MM/YYYY`<br>• `DD/MM/YYYY'T'HH:mm`<br>• `DD/MM/YYYY'T'HH:mm:ss`<br><br> This signifies the end date of analysis. If the field is ignored, the date will be set to the default one or the date indicated in CLI flags <br><br> If hours/ minutes/ seconds are not provided, each will default to `00`   |
```
[the remark this answers is not in the document]
Comment (gabriellegtw) [reply docs/ug/configFiles.md]: Ooo thanks for catching that! I've edited it - hope it's okay now

**Review by CYX22222003 — APPROVED** [review]
LGTM

**Review by Airiinnn — CHANGES_REQUESTED** [review]
Thanks for the PR! 

I think the tests could be refined a little to use boundary values. We want to make sure that the added support for time is also behaving correctly when interacting with the CLI `--since` and `--until` datetimes, which were the point of some of these tests.

Regardless, good job on the change and for adding thorough tests and documentation!

**Comment by gabriellegtw** [comment]
Hi @Airiinnn , thanks so much for the comments! I have changed the test cases to test the boundary values. Do let me know if there's anything else to be improved on 

Regarding the last comment on the `repoCsvConfig_invalidOverridenDatesWithTime` tests, I added them to test for cases if the CSV contains invalid formats. Should I remove the cases I have added for this? Thank you!

**Review by Airiinnn — COMMENTED** [review]
Hi @gabriellegtw , sorry for the late review 🙇 . Thanks again for taking the time to make these changes.

There are actually still some problems with some of the test cases. I think it would be very beneficial for you if you understood what each test you're implementing is checking for. I've tried to explain for some of those that are are not really checking for what it is supposed to. If you need anymore clarification, please feel free to ping us. 

While having comprehensive tests are good, I think it's more important to intentionally test for cases. If you feel that some cases were already covered by other tests, it may also make sense not to repeat them if it's not really checking something new.

But regardless, thanks again for your effort! Sorry this change became a bit draggy. If you need any more help or clarification, just ping us. 🙏

File: src/test/java/reposense/parser/RepoConfigParserTest.java [hunk 4]
```diff
@@ -95,11 +113,16 @@ public class RepoConfigParserTest {
...
+    private static final LocalDateTime TEST_REPO_SINCE_DATE_WITH_TIME = LocalDateTime.of(
             2024, Month.SEPTEMBER, 21, 0, 0, 0);
-    private static final LocalDateTime TEST_ARTIFICIAL_UNTIL_DATE = LocalDateTime.of(
+    private static final LocalDateTime TEST_REPO_UNTIL_DATE_WITH_TIME = LocalDateTime.of(
             2025, Month.JANUARY, 29, 23, 59, 59);
 
+    private static final LocalDateTime TEST_ARTIFICIAL_SINCE_DATE = LocalDateTime.of(
+            2024, Month.SEPTEMBER, 21, 13, 30, 0);
+    private static final LocalDateTime TEST_ARTIFICIAL_UNTIL_DATE = LocalDateTime.of(
+            2025, Month.JANUARY, 29, 17, 59, 59);
```
Comment (Airiinnn) [inline src/test/java/reposense/parser/RepoConfigParserTest.java:124]: Just for extra clarity:

`TEST_REPO_DEFAULT_SINCE_DATE` & `TEST_REPO_DEFAULT_UNTIL_DATE` were used for dates in CSVs. They were intentially selected to match the dates for valid cases of repos in the CSVs.

`TEST_ARTIFICIAL_SINCE_DATE` & `TEST_ARTIFICIAL_UNTIL_DATE` were used for CLI dates. (Used with `InputBuilder` in tests).

If it helps, you could also rename them to make them clearer.

File: src/test/java/reposense/parser/RepoConfigParserTest.java [hunk 5]
```diff
@@ -194,15 +277,81 @@ public void repoCsvConfig_completeCliDatesCsvDatesWithinRange_success() throws E
...
+        Assertions.assertTrue(configCharlie.isHasUpdatedSinceDateInConfig());
+        Assertions.assertTrue(configCharlie.isHasUpdatedUntilDateInConfig());
+        Assertions.assertEquals(configCharlie.getUntilDate(), TEST_ARTIFICIAL_UNTIL_DATE);
+        Assertions.assertEquals(configCharlie.getSinceDate(), TEST_ARTIFICIAL_SINCE_DATE);
+
+        Assertions.assertTrue(configAlpha.isHasUpdatedSinceDateInConfig());
+        Assertions.assertTrue(configAlpha.isHasUpdatedUntilDateInConfig());
+        Assertions.assertEquals(configAlpha.getSinceDate(), TEST_ARTIFICIAL_SINCE_DATE);
+        Assertions.assertEquals(configAlpha.getUntilDate(), TEST_ARTIFICIAL_UNTIL_DATE);
+    }
```
[the remark this answers is not in the document]
Comment (gabriellegtw) [reply src/test/java/reposense/parser/RepoConfigParserTest.java:353]: Hi @Airiinnn 

Thank you so much for the very detailed explanation and for bearing with me as I clarify these points. I really appreciate your patience!

Regarding testing the boundary values, I was thinking that since it might be difficult to differentiate in a single test case when values are from the config and when they are from the CLI, would it be better for me to split them into separate test cases (i.e., one dedicated to testing the boundary scenarios and another for the non-boundary cases)?

Also, I apologize for not raising this question earlier. I actually had the same doubt before submitting the last time, but wasn’t sure how to bring it up. Thanks again for the patience in helping me understand this better!

[the remark this answers is not in the document]
Comment (Airiinnn) [reply src/test/java/reposense/parser/RepoConfigParserTest.java:353]: Hi @gabriellegtw, no worries. You're actually helping us by contributing to the project so we appreciate any effort you put in.

Regarding boundary values, correct me if I misunderstood, but in this test case, I don't think there is a need to split into separate cases?

For this tests, we want to verify that when CLI dates are provided, and all CSV dates are within the the CLI date range, the CSV dates are taken. For missing CSV dates, CLI dates should then be used. 

If we did something like this, we would be testing with boundary values while being able to check missing CSV dates are replaced with CLI dates.

image

`repoCsvConfig_completeCliDatesAndTimesSecondsMixedValidCsv_success()` can follow something similar:

image

(Beta, Charlie invalid, check not included. Alpha valid, check the final range.)

But ultimately, up to you how you want to structure your tests. The important thing is the tests are covering most cases that need to be verified. If you think a certain behavior has been covered by another test, it's fine to not repeat it again.

Apart from these 2 and comments on the old tests, the rests of the tests are fine!

CS2103T reference about boundary values (if interested): https://nus-cs2103-ay2425s1.github.io/website/se-book-adapted/chapters/testCaseDesign.html#boundary-value-analysis

[the remark this answers is not in the document]
Comment (gabriellegtw) [reply src/test/java/reposense/parser/RepoConfigParserTest.java:353]: Hi @Airiinnn ,

Thanks for the very comprehensive diagrams! I think my initial confusion arose because I thought that the start date of Charlie should also be `21/09/2024T13:30:00` (which resulted in me misusing `TEST_ARTIFICIAL_*_DATE`). This comment is really clear :D Will submit a revision soon. Thanks!

**Comment by gabriellegtw** [comment]
Hi @Airiinnn, once again, thank you so much for your patience in explaining the fixes to me — I've just submitted another revision.

Just a quick note on the `TEST_ARTIFICIAL_SINCE_DATE` and `TEST_ARTIFICIAL_UNTIL_DATE` variables, I remember in a previous comment you mentioned the following:

> *“Could we stick to just 1 TEST_ARTIFICIAL_SINCE_DATE and 1 TEST_ARTIFICIAL_UNTIL_DATE? Maybe the ones with time and seconds if they're needed for the new tests...”*

I ended up keeping both the original (`TEST_ARTIFICIAL_SINCE_DATE` / `TEST_ARTIFICIAL_UNTIL_DATE`) and the time-specific versions (`TEST_ARTIFICIAL_SINCE_DATE_WITH_TIME` / `TEST_ARTIFICIAL_UNTIL_DATE_WITH_TIME`) to support some of the older tests that I didn't write. That said, I took out `TEST_ARTIFICIAL_SINCE_DATE_WITH_TIME_SECONDS` / `TEST_ARTIFICIAL_UNTIL_DATE_WITH_TIME_SECONDS`

Thanks!

**Review by Airiinnn — APPROVED** [review]
LGTM! Thanks for your contributions!
